modules/miner/blockmanager.go

```diff
@@ -69,6 +70,9 @@ func (m *Miner) newSourceBlock() {
 // it is typically safe to assume that headers will be remembered for
 // min(10 minutes, 10e3 requests).
 func (m *Miner) HeaderForWork() (types.BlockHeader, types.Target, error) {
+	if m.tg.IsStopped() {
+		build.Critical(errClosed)
+	}
```

Comment:
As Luke pointed out, simpling calling `m.tg.Add` is enough because it will check `IsStopped` inside of that function. Then you can return whatever error it returns.